proposals/11-21-krm-fn-k8s-backend.md

```diff
@@ -0,0 +1,218 @@
+constant fields, or introduce one more extra argument. Since it's desired to minimize the number of arguments, the first option
+with copying labels from template will be selected. User may set the PodTemplate name with `--pod-template-name` argument and the pod
+template must exist in the same namespace where krm-function pod is going to be ran. The user may skip that parameter and by
+default the [following PodTemplate](https://github.com/kubernetes-sigs/kustomize/pull/4260/files#diff-bcc3694e89726203de4e763f2130891ec669833ab26f3856bcddde18c6514a9fR27)
+will be used. Env variables are put to that template already (TBD: maybe it's better to append env vars, rather than just replace the env vars already presented in the
+template). `--mount` flag is ignored as of now in case kubectl-backend is used, but there is a possible option to take what the set of volumes from PodTemplate
+and to append that with data from `--mount` flag - this is a potential topic for discussion.
+
+There is one more potential improvement that should be taken into account: with docker backend krm-function couldn't run
+another krm-functions. For that reason kustomize krm-function was deleted - kustomize couldn't run
+krm-functions without docker-in-docker. With kubectl backend if we could specify service account for the pod
+running krm-function we could run another krm-function initiated by the original krm-function and our
+example with kustomize-krm-function could become possible and security-aspects could be controlled with
+standard k8s-native configuration in the PodTemplate + couple of additional k8s resources.
+This is one more benefit of the approach with `--pod-template-name` flag.
+
+The proposed implementation has one more flag `--pod-start-timeout`. It was added, because depending on the cluster and the image size to pull the
+time for pod start may vary. If the user doesn't set that, the timeout is 60s by default. The previous implementation of RunFn in kyaml
+didn't work with timeouts. In contrast the updated modification that is used in `kpt` has timeouts that are applicable to the whole container execution including
+time needed to pull the image. In general, it's a good practice for automated tools to be able to set timeouts for operations. One potential topic
+for discussion may be - to create a common approach for that, e.g., to set only the overall timeout E.g. if timeout for that particular krm-functions is set
+for the whole execution, this timeout can be set in function config and we may get rid of this argument. That means we may want to stick to the approach
+that `kpt` selected.
+
+We're using the following sequence to run krm-function in the pod:
+1. If PodTemplateName is set - taking it from k8s by kubectl (the same namespace), OR using defaultPodTemplate overwise. See some mandatory fields values there, e.g. stdin,
```

Comment:
Is there a strong reason to support `PodTemplateName` from the beginning?
IMO we can start simply with a static PodTemplate. Support `PodTemplateName` later if needed.

Reply:
several reasons why I proposed that:
1. I wasn't sure if my default template is secure enough be default - so I added the way to override that. I guess it will take some time to figure out what should be in the default template to make it universal enough and secure at the same time.
2. It gives a fine-grain control on what we're doing in template and user can use it to basically do whatever they need. I see a lot of different applications
3. since PodTemplateName has to be set explicitly - the user should understand what he is doing and what will be consequences. Just like with `--enable-exec`
4. it's already implemented in https://github.com/kubernetes-sigs/kustomize/pull/4260

Reply:
this doc already has some info in several places why podTemplate is option I would like to have from the beginning. by default it works with static podTemplate